service/history/api/startworkflow/api.go

```diff
@@ -617,6 +625,79 @@ func (s *Starter) getWorkflowHistory(ctx context.Context, mutableState *mutableS
 	return events, nil
 }
 
+func (s *Starter) handleUseExistingWorkflowOnConflictOptions(
+	ctx context.Context,
+	workflowKey definition.WorkflowKey,
+	currentWorkflowConditionFailed *persistence.CurrentWorkflowConditionFailedError,
+) (*historyservice.StartWorkflowExecutionResponse, StartOutcome, error) {
+	var err error
+	onConflictOptions := s.request.StartRequest.GetOnConflictOptions()
+	if onConflictOptions != nil {
+		requestID := ""
+		if onConflictOptions.AttachRequestId {
+			requestID = s.request.StartRequest.GetRequestId()
+		}
+		var completionCallbacks []*commonpb.Callback
+		if onConflictOptions.AttachCompletionCallbacks {
+			completionCallbacks = s.request.StartRequest.GetCompletionCallbacks()
+		}
+		var links []*commonpb.Link
+		if onConflictOptions.AttachLinks {
+			links = s.request.StartRequest.GetLinks()
+		}
+		err = api.GetAndUpdateWorkflowWithNew(
+			ctx,
+			nil,
+			workflowKey,
+			func(workflowLease api.WorkflowLease) (*api.UpdateWorkflowAction, error) {
+				mutableState := workflowLease.GetMutableState()
+				if !mutableState.IsWorkflowExecutionRunning() {
+					return nil, consts.ErrWorkflowCompleted
+				}
+
+				_, err := mutableState.AddWorkflowExecutionOptionsUpdatedEvent(
+					nil,
+					false,
+					requestID,
+					completionCallbacks,
+					links,
+				)
+				return api.UpdateWorkflowWithoutWorkflowTask, err
+			},
+			nil, // no new workflow
+			s.shardContext,
+			s.workflowConsistencyChecker,
+		)
+	}
+
+	switch err {
+	case nil:
+		resp := &historyservice.StartWorkflowExecutionResponse{
+			RunId:   workflowKey.RunID,
+			Started: false, // set explicitly for emphasis
+		}
+		return resp, StartReused, nil
+	case consts.ErrWorkflowCompleted:
+		// Need to re-evaluate the reuse policy because the previous check didn't lock the current
+		// execution. So it's possible the workflow completed after the first call of
+		// api.ResolveDuplicateWorkflowID and before being able to update the existing workflow.
+		err := api.ResolveWorkflowIDReusePolicy(
+			workflowKey,
+			currentWorkflowConditionFailed.Status,
+			currentWorkflowConditionFailed.RequestID,
+			s.request.StartRequest.GetWorkflowIdReusePolicy(),
+		)
```

Comment:
Is `currentWorkflowConditionFailed.RequestID` the current request's ID? Should we be checking that the request IDs in the DB and request matches? (asking for lack of context on this part of the code).

Reply:
`currentWorkflowConditionFailed.RequestID` is the existing current workflow, which can be running or completed already.